### PR TITLE
[dag] async message handler support

### PR DESCRIPTION
**Pull request by ibalajiarun**
### Description

This PR makes RpcHandler trait handler method async so any implementations can call other async methods. This is necessary to support pulling payload in Dag Driver.

### CODE REVIEW

**Comment by ibalajiarun** [comment]
Current dependencies on/for this PR:
* main
  * **PR #9830** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/9830" target="_blank">Graphite</a> 
    * **PR #9831** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/9831" target="_blank">Graphite</a> 
      * **PR #9724** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/9724" target="_blank">Graphite</a> 
        * **PR #9968** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/9968" target="_blank">Graphite</a> 
          * **PR #9967** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/9967" target="_blank">Graphite</a> 
            * **PR #10006** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/10006" target="_blank">Graphite</a> 
              * **PR #9994** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/9994" target="_blank">Graphite</a> 
                * **PR #10040** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/10040" target="_blank">Graphite</a> 
                  * **PR #10041** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/10041" target="_blank">Graphite</a> 
                    * **PR #10042** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/10042" target="_blank">Graphite</a> 
                      * **PR #10055** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/10055" target="_blank">Graphite</a>  👈
                      * **PR #10007** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/10007" target="_blank">Graphite</a> 
        * **PR #9832** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/9832" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/aptos-labs/aptos-core/10055?utm_source=stack-comment).

File: consensus/src/dag/dag_driver.rs [hunk 1]
```diff
@@ -96,37 +100,62 @@ impl DagDriver {
...
+        debug!("entering new round {}", new_round);
         // TODO: support pulling payload
-        let payload = Payload::empty(false);
+        let payload = match self
+            .payload_client
+            .pull_payload(
+                Duration::from_secs(1),
+                100,
+                1000,
+                PayloadFilter::Empty,
```
Comment (sasha8) [inline consensus/src/dag/dag_driver.rs:145]: We need a filter. 
I think we forgot about this functionality, we need to maintain a data structure with all the batches currently in local DAG. (Or we can traverse the dag each time before pulling.)

File: consensus/src/dag/dag_handler.rs [hunk 2]
```diff
@@ -102,32 +128,30 @@ impl NetworkHandler {
...
         }
 
-        let response: anyhow::Result<DAGMessage> = match dag_message {
-            DAGMessage::NodeMsg(node) => node
-                .verify(&self.epoch_state.verifier)
-                .and_then(|_| self.node_receiver.process(node))
-                .map(|r| r.into()),
+        let response: anyhow::Result<DAGMessage> = {
+            match self.verify_incoming_rpc(&dag_message) {
+                Ok(_) => match dag_message {
```
Comment (sasha8) [inline consensus/src/dag/dag_handler.rs:133]: nit: you can combine these two matches by returning Result<DAGMessage>, but not sure which will be more readable.

**Review by sasha8 — COMMENTED** [review]
We could potentially block_on pull_payload in enter_new_round instead of the async trait, right?

File: consensus/src/dag/dag_driver.rs (hunk 1 above)
Comment (ibalajiarun) [reply consensus/src/dag/dag_driver.rs:145]: yes, need to add a TODO. @zekun000  is taking care of it.

**Comment by ibalajiarun** [comment]
> We could potentially block_on pull_payload in enter_new_round instead of the async trait, right?

I think that would block the bootstrapper and not drive other futures in the select! like shutdown call until pull returns.

File: consensus/src/dag/dag_handler.rs [hunk 3]
```diff
@@ -76,19 +76,45 @@ impl NetworkHandler {
...
+            DAGMessage::CertifiedNodeMsg(certified_node) => {
+                certified_node.verify(&self.epoch_state.verifier)
+            },
+            DAGMessage::FetchRequest(request) => request.verify(&self.epoch_state.verifier),
+            _ => {
+                error!(
+                    "unknown rpc message {:?}",
+                    std::mem::discriminant(dag_message)
+                );
+                Err(anyhow::anyhow!("unexpected rpc message"))
```
Comment (zekun000) [inline consensus/src/dag/dag_handler.rs:113]: we should avoid double logging, either we return error or we log and return nothing. I think for function like this we can just return the error back but in process_rpc should probably just log it and only propagate NeedSync status

**Review by zekun000 — APPROVED** [review]

File: consensus/src/dag/dag_handler.rs (hunk 2 above)
Comment (ibalajiarun) [reply consensus/src/dag/dag_handler.rs:131]: I am still thinking how better to rewrite this fn, but not in this PR.

**Review by sasha8 — APPROVED** [review]

**Pull request merged**